platforms/documentation/docs/src/docs/userguide/dep-man/basics/variant_aware_resolution.adoc

```diff
@@ -0,0 +1,142 @@
+This makes it available for other projects to consume.
+
+====
+include::sample[dir="snippets/dependencyManagement/modelingFeatures-crossProjectPublications-advanced/kotlin",files="producer/build.gradle.kts[tags=attach-outgoing-artifact]"]
+include::sample[dir="snippets/dependencyManagement/modelingFeatures-crossProjectPublications-advanced/groovy",files="producer/build.gradle[tags=attach-outgoing-artifact]"]
+====
+
+What we have done here is that we have added a _new_ *variant*, which can be used _at runtime_, but contains instrumented classes instead of the normal classes.
+However, it now means that for runtime, the consumer has to choose between two variants:
+
+1. `runtimeElements`, the regular variant offered by the `java-library` plugin
+2. `instrumentedJars`, the variant we have created
+
+== The consumer side
+
+=== 1. Add dependencies:
+
+First, on the consumer side, like any other project, we define the Java library as a dependency:
+
+====
+include::sample[dir="snippets/dependencyManagement/modelingFeatures-crossProjectPublications-advanced/kotlin",files="consumer/build.gradle.kts[tags=test_dependency]"]
+include::sample[dir="snippets/dependencyManagement/modelingFeatures-crossProjectPublications-advanced/groovy",files="consumer/build.gradle[tags=test_dependency]"]
+====
+
+At this point, Gradle will still select the default `runtimeElements` *variant* for your dependencies.
+This is because the `testRuntimeClasspath` configuration is requesting *artifacts* with the `jar` library elements *attribute*, while the producer defines the `instrumentedJars` *variant* with a different *attribute*.
+
+=== 2. Adjust the requested attributes:
+
+The `testRuntimeClasspath` configuration is modified to ask for "instrumented-jar" versions of the dependencies.
+This means that when Gradle resolves dependencies for this configuration, it will prefer JAR files that are marked as "instrumented":
+
+====
+include::sample[dir="snippets/dependencyManagement/modelingFeatures-crossProjectPublications-advanced/kotlin",files="consumer/build.gradle.kts[tags=ask-for-instrumented-classes]"]
+include::sample[dir="snippets/dependencyManagement/modelingFeatures-crossProjectPublications-advanced/groovy",files="consumer/build.gradle[tags=ask-for-instrumented-classes]"]
+====
+
+We can look at a report on the consumer side to view exactly what *attributes* of each dependency will be requested:
+
+[source,text]
+----
+$ .gradle resolvableConfigurations --configuration testRuntimeClasspath
+
+Attributes
+    - org.gradle.category            = library
+    - org.gradle.dependency.bundling = external
+    - org.gradle.jvm.version         = 11
+    - org.gradle.libraryelements     = instrumented-jar
+    - org.gradle.usage               = java-runtime
+----
+
+In this case, we see the `testRuntimeClasspath` configuration requests _instrumented classes_.
+
+=== 3. Handle missing variants with compatibility rules:
```

Comment:
Again, this is an advanced topic. I'd vote to omit this here.

Reply:
Removed.